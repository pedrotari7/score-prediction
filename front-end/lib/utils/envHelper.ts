export const backendUrl =
	process.env.NEXT_PUBLIC_APP_ENV === 'local-dev'
		? 'http://localhost:5001/score-prediciton/europe-west1/api/'
		: 'https://europe-west1-score-prediciton.cloudfunctions.net/api';
