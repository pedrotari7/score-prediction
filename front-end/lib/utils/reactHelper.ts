import { DateTime } from 'luxon';

export const classNames = (...classes: any[]) => classes.filter(Boolean).join(' ');

export const getCurrentDate = () => {
	return DateTime.now();
	// Mocked date
	// return DateTime.fromISO('2021-06-16T19:00:00+0000');
};
