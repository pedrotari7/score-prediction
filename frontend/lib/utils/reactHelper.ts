import { DateTime } from 'luxon';
import { Fixture, Prediction, Result, UserResult } from '../../../interfaces/main';

export const classNames = (...classes: string[]) => classes.filter(Boolean).join(' ');

export const getCurrentDate = () => {
	return DateTime.now();
	// Mocked date
	// return DateTime.fromISO('2021-06-16T19:00:00+0000');
	// return DateTime.fromISO('2016-06-09T19:00:00+0000');
};

export const formatScore = (goal: number | null) => {
	if (typeof goal === 'number' && goal >= 0) return goal;
	if (typeof goal === 'number' && goal < 0) return 'H';
	return 'X';
};

export const isNum = (n: number | null) => typeof n === 'number';

export const getOutcome = (g: Result): string | null => {
	if (!isNum(g.home) || !isNum(g.away)) return null;
	if (g.home > g.away) return 'winH';
	if (g.home < g.away) return 'winA';
	if (g.home === g.away) return 'draw';
	return null;
};

export const getResult = (prediction: Prediction, result: Result): Partial<UserResult> => {
	const { home: predH, away: predA } = prediction;
	const { home: realH, away: realA } = result;

	const isExactScore = predH === realH && predA === realA;

	if (isExactScore) return { points: 5, exact: 1 };

	const isCorrectResult =
		!isExactScore && getOutcome(prediction) !== null && getOutcome(prediction) === getOutcome(result);

	if (isCorrectResult) return { points: 3, result: 1 };

	const isCorrectGoal = !isExactScore && !isCorrectResult && (predH === realH || predA === realA);

	if (isCorrectGoal) return { points: 1, onescore: 1 };

	return { points: 0, fail: 1 };
};

export const isGameFinished = (game: Fixture) => ['FT', 'AET', 'PEN'].includes(game.fixture.status.short);
