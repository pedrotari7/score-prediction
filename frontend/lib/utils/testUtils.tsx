import { render } from '@testing-library/react';
import { ReactNode, ContextType } from 'react';
import RouteContext from '../../context/RouteContext';
import NoSpoilersContext from '../../context/NoSpoilersContext';

export const customRender = (
	ui: ReactNode,
	options?: {
		routeContextProps?: ContextType<typeof RouteContext>;
		noSpoilersContextProps?: ContextType<typeof NoSpoilersContext>;
		renderOptions?: Parameters<typeof render>[1];
	}
) => {
	return render(
		<RouteContext.Provider value={options?.routeContextProps ?? null}>
			<NoSpoilersContext.Provider value={options?.noSpoilersContextProps ?? null}>
				{ui}
			</NoSpoilersContext.Provider>
		</RouteContext.Provider>,
		options?.renderOptions
	);
};
