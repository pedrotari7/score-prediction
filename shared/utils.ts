import {
	Competitions,
	Fixture,
	Fixtures,
	Prediction,
	Predictions,
	PredResult,
	Result,
	UserResult,
} from '../interfaces/main';

export const competitions: Competitions = {
	euro2016: { name: 'euro2016', league: 4, season: 2016, start: '2016-06-10', end: '2016-07-10', logo: '/logo.svg' },
	euro2020: { name: 'euro2020', league: 4, season: 2020, start: '2021-06-09', end: '2021-07-15', logo: '/logo.svg' },

	// wc2010: { name: 'wc2010', league: 1, season: 2010, start: '2010-06-10', end: '2010-07-12', logo: '/logo.svg' },
	// wc2014: { name: 'wc2014', league: 1, season: 2014, start: '2014-06-11', end: '2014-07-15', logo: '/logo.svg' },
	// wc2018: { name: 'wc2018', league: 1, season: 2018, start: '2018-06-13', end: '2018-07-16', logo: '/logo.svg' },

	wc2022: {
		name: 'wc2022',
		league: 1,
		season: 2022,
		start: '2022-11-20',
		end: '2022-12-18',
		logo: '/wc2022-logo.svg',
	},
};

export const isNum = (n: number | null): n is number => typeof n === 'number';

export const isEmpty = (obj: Object) => Object.keys(obj).length == 0;

export const getOutcome = (g: Result): 'winH' | 'winA' | 'draw' | null => {
	if (!isNum(g.home) || !isNum(g.away)) return null;
	if (g.home > g.away) return 'winH';
	if (g.home < g.away) return 'winA';
	if (g.home === g.away) return 'draw';
	return null;
};

export const getExtraTimeResult = ({ score: { fulltime, extratime }, fixture, goals }: Fixture) => {
	if (['PEN', 'AET'].includes(fixture.status.short)) {
		return { home: fulltime.home + extratime.home, away: fulltime.away + extratime.away };
	}
	return goals;
};

export const joinResults = (a: Partial<UserResult>, b: Partial<UserResult>) => ({
	...a,
	...b,
	points: (a.points ?? 0) + (b.points ?? 0),
});

export const getResult = (prediction: Prediction, game: Fixture): Partial<UserResult> => {
	const result = getExtraTimeResult(game);

	let ret = {};

	const { home: predH, away: predA } = prediction;
	const { home: realH, away: realA } = result;

	const isExactScore = predH === realH && predA === realA;

	if (isExactScore) ret = joinResults(ret, { points: 5, exact: 1 });

	const isCorrectResult =
		!isExactScore && getOutcome(prediction) !== null && getOutcome(prediction) === getOutcome(result);

	if (isCorrectResult) ret = joinResults(ret, { points: 3, result: 1 });

	const isCorrectGoal = !isExactScore && !isCorrectResult && (predH === realH || predA === realA);

	if (isCorrectGoal) ret = joinResults(ret, { points: 1, onescore: 1 });

	const isPenaltyWinner =
		isPenaltyShootout(game) &&
		getOutcome(prediction) !== null &&
		getOutcome(prediction) === getOutcome(game.score.penalty);

	if (isPenaltyWinner) ret = joinResults(ret, { points: 1, penalty: 1 });

	if (isEmpty(ret)) return { points: 0, fail: 1 };

	return ret;
};

export const isGameFinished = (game: Fixture) => ['FT', 'AET', 'PEN', 'INT', 'PST'].includes(game.fixture.status.short);

export const isPenaltyShootout = (game: Fixture) => game.fixture.status.short === 'PEN';

export const isGameStarted = (game: Fixture) => game.fixture.status.short !== 'NS';

export const isGameOnGoing = (game: Fixture) => !isGameFinished(game) && isGameStarted(game);

export const initializeTeamResult = (): PredResult => ({
	points: 0,
	wins: 0,
	draws: 0,
	loses: 0,
	ga: 0,
	gc: 0,
});

export const calculatePoints = ({ wins, draws }: PredResult) => 3 * wins + draws;

export const calculateResults = (fixtures: Fixture[], predictions: Predictions, uid: string) => {
	const isGroupStage = (f: Fixture) => f.league.round.includes('Group');

	return fixtures.filter(isGroupStage).reduce((teams, game) => {
		const homeTeam = game.teams.home.id;
		const awayTeam = game.teams.away.id;

		if (!(homeTeam in teams)) teams[homeTeam] = initializeTeamResult();
		if (!(awayTeam in teams)) teams[awayTeam] = initializeTeamResult();

		const prediction = predictions?.[game.fixture.id]?.[uid];

		if (prediction?.home >= 0 && prediction?.away >= 0) {
			if (prediction.home > prediction.away) {
				teams[homeTeam].wins += 1;
				teams[awayTeam].loses += 1;
			} else if (prediction?.home < prediction?.away) {
				teams[awayTeam].wins += 1;
				teams[homeTeam].loses += 1;
			} else {
				teams[homeTeam].draws += 1;
				teams[awayTeam].draws += 1;
			}
			teams[homeTeam].points = calculatePoints(teams[homeTeam]);
			teams[awayTeam].points = calculatePoints(teams[awayTeam]);

			teams[homeTeam].ga += prediction?.home ?? 0;
			teams[homeTeam].gc += prediction?.away ?? 0;
			teams[awayTeam].ga += prediction?.away ?? 0;
			teams[awayTeam].gc += prediction?.home ?? 0;
		}

		return teams;
	}, {} as Record<number, PredResult>);
};

export const compareResults = (rA: PredResult, rB: PredResult) => {
	if (rA.points !== rB.points) return rB.points - rA.points;

	const gdA = rA.ga - rA.gc;
	const gdB = rB.ga - rB.gc;

	if (gdA !== gdB) return gdB - gdA;

	if (rA.ga !== rB.ga) return rB.ga - rA.ga;
	return 0;
};

export const sortGroup = (
	groupTeams: number[],
	teamsResults: Record<number, PredResult>,
	fixtures: Fixtures,
	predictions: Predictions,
	userID: string
) => {
	const teamsWithSamePoints = groupTeams.reduce<Record<number, number[]>>((pack, teamID) => {
		const points = teamsResults[teamID].points;
		if (!(points in pack)) pack[points] = [];
		pack[points].push(teamID);
		return pack;
	}, {});

	const sortedStandings = Object.entries(teamsWithSamePoints)
		.sort(([a], [b]) => parseInt(b) - parseInt(a))
		.flatMap(([_, teams]) => {
			if (teams.length === 1) return teams;

			const tiedTeamsGames = Object.values(fixtures).filter(
				(f: Fixture) => teams.includes(f.teams.home.id) && teams.includes(f.teams.away.id)
			);

			const tiedTeamsResults = calculateResults(tiedTeamsGames, predictions, userID);

			const sortedTeams = Object.entries(tiedTeamsResults)
				.sort(([teamA, resultsA], [teamB, resultsB]) => {
					const localResult = compareResults(resultsA, resultsB);

					if (localResult) return localResult;

					return compareResults(teamsResults[parseInt(teamA)], teamsResults[parseInt(teamB)]);
				})
				.map(([t]) => parseInt(t));

			return sortedTeams;
		});

	return sortedStandings;
};
