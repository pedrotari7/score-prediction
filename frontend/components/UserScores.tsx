import React, { ReactNode } from 'react';
import { User } from '../../interfaces/main';
import { classNames } from '../lib/utils/reactHelper';
import { Tooltip } from 'react-tooltip';

export const Circle = ({ children, className }: { children: ReactNode; className: string }) => (
	<div
		className={classNames(
			'mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center',
			className
		)}
	>
		{children}
	</div>
);

const TooltipInto = ({ children, content }: { children: ReactNode; content: string }) => (
	<a data-tooltip-id='my-tooltip' data-tooltip-content={content} data-tooltip-place='bottom'>
		{children}
	</a>
);

export const UserScores = ({ user, stage }: { user: User; stage: string }) => {
	return (
		<div className='flex flex-row flex-wrap items-center justify-center'>
			<Tooltip id='my-tooltip' />
			<TooltipInto content='exact'>
				<Circle className='bg-green-600'>{user.score[stage].exact}</Circle>
			</TooltipInto>
			<TooltipInto content='result'>
				<Circle className='bg-yellow-600'>{user.score[stage].result}</Circle>
			</TooltipInto>

			<TooltipInto content='onescore'>
				<Circle className='bg-pink-600'>{user.score[stage].onescore}</Circle>
			</TooltipInto>

			<TooltipInto content='fail'>
				<Circle className='bg-red-600'>{user.score[stage].fail}</Circle>
			</TooltipInto>

			<TooltipInto content='groups'>
				<Circle className='bg-purple-700'>{user.score[stage].groups}</Circle>
			</TooltipInto>

			<TooltipInto content='penalty'>
				<Circle className='bg-gray-500'>{user.score[stage].penalty}</Circle>
			</TooltipInto>

			<TooltipInto content='points'>
				<Circle className='my-2 size-10 bg-gray-700 p-4'>{user.score[stage].points}</Circle>
			</TooltipInto>
		</div>
	);
};
