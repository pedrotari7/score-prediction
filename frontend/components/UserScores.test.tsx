import type { User } from '../../interfaces/main';
import { customRender } from '../lib/utils/testUtils';
import { Circle, UserScores } from './UserScores';

jest.mock('react-tooltip');

describe('UserScores', () => {
	const user = {
		score: {
			all: {
				exact: 1,
				onescore: 2,
				points: 3,
				result: 4,
				penalty: 5,
				fail: 6,
				groups: 7,
			},
			Groups: {
				exact: 8,
				onescore: 9,
				points: 10,
				result: 11,
				penalty: 12,
				fail: 13,
				groups: 14,
			},
		},
	} as Partial<User> as User;

	it('renders user scores with a stage all', () => {
		const { container } = customRender(<UserScores user={user} stage='all' />);
		expect(container).toMatchSnapshot();
	});

	it('renders user scores with a stage groups', () => {
		const { container } = customRender(<UserScores user={user} stage='Groups' />);
		expect(container).toMatchSnapshot();
	});
});

describe('Circle', () => {
	it('renders circle with children ', () => {
		const { container } = customRender(
			<Circle className='test'>
				<div>test children</div>
			</Circle>
		);
		expect(container).toMatchSnapshot();
	});
});
