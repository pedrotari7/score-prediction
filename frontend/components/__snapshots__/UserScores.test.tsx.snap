// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle renders circle with children  1`] = `
<div>
  <div
    class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center test"
  >
    <div>
      test children
    </div>
  </div>
</div>
`;

exports[`UserScores renders user scores with a stage all 1`] = `
<div>
  <div
    class="flex flex-row flex-wrap items-center justify-center"
  >
    <a
      data-tooltip-content="exact"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-green-600"
      >
        1
      </div>
    </a>
    <a
      data-tooltip-content="result"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-yellow-600"
      >
        4
      </div>
    </a>
    <a
      data-tooltip-content="onescore"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-pink-600"
      >
        2
      </div>
    </a>
    <a
      data-tooltip-content="fail"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-red-600"
      >
        6
      </div>
    </a>
    <a
      data-tooltip-content="groups"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-purple-700"
      >
        7
      </div>
    </a>
    <a
      data-tooltip-content="penalty"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-gray-500"
      >
        5
      </div>
    </a>
    <a
      data-tooltip-content="points"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center my-2 size-10 bg-gray-700 p-4"
      >
        3
      </div>
    </a>
  </div>
</div>
`;

exports[`UserScores renders user scores with a stage groups 1`] = `
<div>
  <div
    class="flex flex-row flex-wrap items-center justify-center"
  >
    <a
      data-tooltip-content="exact"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-green-600"
      >
        8
      </div>
    </a>
    <a
      data-tooltip-content="result"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-yellow-600"
      >
        11
      </div>
    </a>
    <a
      data-tooltip-content="onescore"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-pink-600"
      >
        9
      </div>
    </a>
    <a
      data-tooltip-content="fail"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-red-600"
      >
        13
      </div>
    </a>
    <a
      data-tooltip-content="groups"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-purple-700"
      >
        14
      </div>
    </a>
    <a
      data-tooltip-content="penalty"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center bg-gray-500"
      >
        12
      </div>
    </a>
    <a
      data-tooltip-content="points"
      data-tooltip-id="my-tooltip"
      data-tooltip-place="bottom"
    >
      <div
        class="mx-1 flex size-6 select-none flex-row items-center justify-center rounded-full p-4 text-center my-2 size-10 bg-gray-700 p-4"
      >
        10
      </div>
    </a>
  </div>
</div>
`;
